### DISCUSSION TITLE
Rethinking Static Variables In V

### DISCUSSION THREAD

**Discussion by Wajinn** [vlang/v#13926] [General]
In looking at static variables, I'm wondering if their usefulness is being underestimated, particularly in a non class-based OOP language that emphasizes the use of functions.  Perhaps static variables should not be designated as "unsafe", and be allowed to play a more important role in V.  Additionally, static variables should not be confused with global variables (particularly in the conceptual sense and usage).

Since V doesn't allow global variables by default, this creates a problem.  There can be a need to **preserve values**, which **changes** (is mutable) based on being acted upon.  What static variables allow, is reducing the scope of such interactions to a **particular function**.  This becomes more important in a language like V, where it doesn't uses class-based OOP nor global variables (normally).

Another area where this issue shows itself is with GUIs.  If you have a function that contains a GUI (it is better to isolate the UI from business logic), often you will need UI related variables that maintain track of state between function calls.  It is arguably much more convenient to use static variables to do this, than workarounds.

What say you all?

**Comment by JalonSolov** [comment]
In V, maintaining state/tracking variables is usually done in structs, with work being done by receiver methods on those structs.  There's no real need to expand static variables to more global usage, when the struct + receiver methods idea is there.

**Reply by Wajinn** [reply]
> There's no real need to expand static variables to more global usage...

But why do static variables have to be designated as "unsafe"?

The line of reasoning as to where I'm going is not that static variables should be elevated to becoming the equivalent to global variables.  I agree with V's core concept of normally disallowing global variables, unless under very special circumstances or lower level coding (thus compiler flag).  I agree that global variables are usually (but not always) bad, in the context of any function being allowed to modify it.  But, I do think that static variables should be looked at separately and distinctly, because of V's emphasis on functions.  The static variable only being able to be modified by the function that contains it.

As the static variable is used within a function, then arguably state/tracking can be completely controlled by the function that contains it without resorting to any mechanism to prevent access as can be the case with a struct.  I state this, because it is possible for GUIs to be contained completely within a function and all necessary state/tracking to also be done inside of the function as well.  The function becomes primary, and there can be greater clarity and separation of UI and business logic.

**Comment by trufae** [comment]
Global variables introduce several problems. Initialization can introduce undesired bugs:
- linking order specifies the constructor order which can result on uaf
- using library constructors and destructors dont play well with multithreading
- Android library linking only runs the constructor once. Which is a problem if those globals are modified at any time. Globals are mutable consts..
- Race conditions if not protected properly using sync getter/setter methods.
- They are bad for libraries, can be handy for single threaded programs. But doesnt scales

Those are the top items in my head right now, considering global variables require more dedication than local ones and are considered a bad practice i think it's better to keep enforcing the encapsulation inside structs


If you have a specific case to share i think it will be good to discuss it and find a solution but in my life as developer i have encountered many bugs because of using global variables in c, c++, java..

**Reply by Wajinn** [reply]
> Global variables introduce several problems. Initialization can introduce undesired bugs:

Static variables are not global variables.  More so in the context of the user, though the distinction between them can be can be less so (though still present) at the lower levels.

In looking at other discussions and issues, it appears as the concept behind the two are confused or at least there are conflicting definitions where they are mostly merged as the same thing.

A global variable can usually be modified by any function. 

A static variable is inside of a function, and changes to the static variable only happen as a result of calling the function that contains it.  How the static variable is modified is controlled by the function that contains it, and any function can not modify it directly.

Though global variables are villainized (and some are good reasons), why can't the distinction in usage allow for greater leniency in the case of static variables used only within a function?

**Reply by ntrel** [reply]
> i think it's better to keep enforcing the encapsulation inside structs

static local variables actually have far tighter encapsulation than a struct passed to a function.

**Comment by Wajinn** [comment]
Arguably the closest "workaround" to static variables (since they are designated as unsafe in V), would be to use closures. An inner nested function can access the variable of the outer function that contains it.  However, closures in V are not finished for all the supported OSes.

**Comment by Wajinn** [comment]
Let's compare them together, with a basic example, so it would be easier to visualize what is being discussed.  As JalonSolov stated, we could use a struct, and there is nothing wrong with that way.  But, this doesn't appear to be intuitive or mentioned in the documentation. 

In contrast, we could put this variable state/tracking in a function, as a static variable.  However, we are forced to type "unsafe" in multiple places, which is aesthetically disturbing and gives the impression we are doing something "dirty", at the least.

If static variables are of use in V, it brings along the questions as why must they come with "warning labels" and have a negative connotation, when they provide greater encapsulation (as mentioned by ntrel) and other potential benefits or ease of use.  This also brings up what is the criteria for memory-unsafe, as it pertains to static variables.  It's at least good to point this out.

```
struct Gadget {
	mut:
	x int = 1
}

fn main() {
	mut struct_adding := Gadget{}

	println(struct_adding.add())
	println(struct_adding.add())

	println(unsafe {foo_add()})
	println(unsafe {foo_add()})
}

fn(mut adder Gadget) add() int {
	adder.x += 1
	return adder.x
}

[unsafe]
fn foo_add() int {
	mut static x := 1
	x += 1
	return x
}
```

**Reply by JalonSolov** [reply]
> But, this doesn't appear to be intuitive or mentioned in the documentation.

Then we just need to improve the documentation.

> However, we are forced to type "unsafe" in multiple places, which is aesthetically disturbing and gives the impression we are doing something "dirty", at the least.

It is.  :-)  You are trying to do something which V discourages.

> If static variables are of use in V, it brings along the questions as why must they come with "warning labels" and have a negative connotation, when they provide greater encapsulation (as mentioned by ntrel) and other potential benefits or ease of use. This also brings up what is the criteria for memory-unsafe, as it pertains to static variables. It's at least good to point this out.

Static variables were really only added for `vinix`, and mainly because they are simpler/more efficient in an OS than using structs.

They are available if you absolutely need them for performance, but are discouraged for general use, mainly because they have "silent" side-effects which can confuse people.

Here's one particular reason NOT to use static variables:

"`static` makes code untestable, because `static` variables tend to be non-replaceable dependencies. Or in more OOP-y words: you aren't following the Dependency Inversion Principle. I came to C and C++ from dynamic languages such as Perl, so my cost model is slanted towards virtual dispatch and function pointers and so on. With current languages, there's some conflict between testability and good architecture, but I think the minor nuisance of making your dependencies explicit and letting them be overridden in tests is noticeably offset by the ease of writing tests, and thus making sure your software is working as expected. Without making your code more dynamic, the only available mechanism to inject dependencies for a test is conditional compilation."

**Reply by ntrel** [reply]
> You are trying to do something which V discourages

Discouraging something is different from memory safety. `unsafe` should only apply to code that is not checked by the compiler for safety.

> static variables tend to be non-replaceable dependencies

This is a good point (but not related to memory safety). However how do you test that something was printed with `println`? You can't pass a mut variable instead of stdout because there's no parameter for that. So should println require `unsafe` for consistency?

**Reply by Wajinn** [reply]
> Then we just need to improve the documentation.

Definitely agree that the documentation should be updated to reflect team V's preferred direction.  Variable state/tracking is clearly an important issue for many programmers.  To include, clarity in separating GUIs from business logic. Perhaps show the alternatives to local static variables with the struct + receiver methods and possibly also closures (another alternative).

In addition to the memory safety point brought up by ntrel, the preference for struct + receiver for variable state/tracking might be more fully understood if it was having some connection to memory management (Autofree or GC).  That is, struct + receiver is related to team V's strategy towards automatic memory management.  Thus what is memory-unsafe can also be referring to that context.  Otherwise, the preference for a little less intuitive (and undocumented) way of variable state/tracking (struct + receiver) can be a bit harder to understand, since static variables were "good enough" to be used in Vinix (great project).
